### ISSUE TITLE
notify_airbrake persists user login during 422 error

### ISSUE THREAD

**Issue by thejourneydude** [airbrake/airbrake#641]
I'm currently handling how my application handles 422 errors (invalid authenticity token). Right now I have this...

```
class ApplicationController < ActionController::Base
  protect_from_forgery with: :exception
  rescue_from ActionController::InvalidAuthenticityToken,
              with: :rescue_422_error

  # This method is called whenever a CSRF token is invalid.
  def rescue_422_error(exception)
    # Since InvalidAuthenticityToken is not raised,
    # we need to notify Airbrake that this error has occurred.
    notify_airbrake(exception)
    redirect_to '/422'
  end
end
```

Whenever I log in with an invalid authenticity token, the user is logged in and is redirected to the /422 page. This is something I don't want to happen. No user should be logged in when they have an invalid authenticity token. What I've realized is that if I comment out `notify_airbrake(exception)`, the user will not be logged in. I believe that `notify_airbrake` is persisting the request and allowing the user to be logged in. I'm currently using Airbrake (5.3.0) and Devise (3.4.1). I have written no custom code for Devise.

**Comment by kyrylo** [comment]
How do you implement your `current_user` method?

**Comment by thejourneydude** [comment]
Can you elaborate on that question?

**Comment by kyrylo** [comment]
So far I don't understand how Airbrake could log in the user. The only way we interact with the user model when Devise is present is through Warden: https://github.com/airbrake/airbrake/blob/baf02e0b3b112c34f31880d998344141729bc2ce/lib/airbrake/rack/user.rb#L14

It might be the case that that call does something behind the scenes, but [we also call `current_user`](https://github.com/airbrake/airbrake/blob/baf02e0b3b112c34f31880d998344141729bc2ce/lib/airbrake/rack/user.rb#L28), which is exposed to your app and you can [re]define it like you wish.

I'm trying to gather more information.

**Comment by thejourneydude** [comment]
I'm going post my SessionsController below. `current_user` is used quite a bit throughout the app. That said we do zero customization on the `current_user` method. We are using the Devise method out-of-the-box.

```
class Users::SessionsController < Devise::SessionsController
# before_filter :configure_sign_in_params, only: [:create]
  respond_to :json

  # DELETE /resource/sign_out
  def destroy
    user_id = current_user.id
    super
    session[:login_status] = 'JustLoggedOut'
    session.delete :login_provider 
  end
end
```

**UPDATE**
I've inspected `current_user` *AFTER* `notify_airbrake` is called and am able to see the contents of the `current_user`. If I remove `notify_airbrake` but still call `current_user`, the user will stay logged in. I must now suspect the act of calling `current_user` is causing this issue. I'm going to close this issue and inspect this further. Thanks for the help.

**Comment by kyrylo** [comment]
Sounds good. I'd be curious to read about your investigations.

**Comment by thejourneydude** [comment]
@kyrylo From what I've gathered so far, Devise/Warden will attempt to authenticate the user when calling `current_user`. This is the source of my problem.

My hack solution is to run `warden.authenticated?` before Airbrake calls `current_user`. I'm honestly not sure if this is code I would use in production but at least is provides one possible solution to my problem.

The code I've written to get this to work is below.

https://github.com/thejourneydude/airbrake/blob/master/lib/airbrake/rack/user.rb#L32

In case anyone wants to see the how Devise/Warden does this, the relevant files are below.

https://github.com/plataformatec/devise/blob/v3.4.1/lib/devise/controllers/helpers.rb#L119
https://github.com/hassox/warden/blob/master/lib/warden/proxy.rb#L103
https://github.com/hassox/warden/blob/master/lib/warden/proxy.rb#L316

**Comment by kyrylo** [comment]
Excellent investigation, thanks! Why do you say you wouldn't use that code in production? Currently, I don't see what's wrong with it and I suggest to open a pull request. Authentication on `notify_airbrake` sounds like a horrible side effect.

**Comment by thejourneydude** [comment]
Thanks! I usually feel pessimistic about code I write. I'll go ahead and test the Airbrake tests on my branch and create a pull request.

**Comment by kyrylo** [comment]
> I usually feel pessimistic about code I write

That may be true, but what's more true is the bug you've found, which isn't trivial at all.
Thanks for posting, it was fixed only with your help :)

**Comment by thejourneydude** [comment]
Thanks for helping me out and quick PR and merge!

**Comment by kyrylo** [comment]
I just released a new version of this gem (v5.7.0) which includes the fix.
